python/pywatchman/__init__.py

```diff
@@ -114,16 +126,128 @@ def __init__(self):
 
     LocalFree = ctypes.windll.kernel32.LocalFree
 
-    GetOverlappedResultEx = ctypes.windll.kernel32.GetOverlappedResultEx
-    GetOverlappedResultEx.argtypes = [wintypes.HANDLE,
-                                      ctypes.POINTER(OVERLAPPED), LPDWORD,
-                                      wintypes.DWORD, wintypes.BOOL]
-    GetOverlappedResultEx.restype = wintypes.BOOL
+    # Windows 7 and earlier does not support GetOverlappedResultEx
+    # The alternative is to use GetOverlappedResult and wait
+    # for read or write operation to complete. This is
+    # done be using CreateEvent and WaitForSingleObject.
+    # Windows 8 is version 6.2 - see the Operating System Version
+    # page on MSDN
+    WINDOWS_8_AND_LATER = sys.getwindowsversion() >= (6, 2)
+    if not WINDOWS_8_AND_LATER:
+        CreateEvent = ctypes.windll.kernel32.CreateEventA
+        CreateEvent.argtypes = [LPDWORD, wintypes.BOOL, wintypes.BOOL,
+                                wintypes.LPSTR]
+        CreateEvent.restype = wintypes.HANDLE
+
+        GetOverlappedResult = ctypes.windll.kernel32.GetOverlappedResult
+        GetOverlappedResult.argtypes = [wintypes.HANDLE,
+                                        ctypes.POINTER(OVERLAPPED), LPDWORD,
+                                        wintypes.BOOL]
+        GetOverlappedResult.restype = wintypes.BOOL
+
+        WaitForSingleObject = ctypes.windll.kernel32.WaitForSingleObject
+        WaitForSingleObject.argtypes = [wintypes.HANDLE, wintypes.DWORD]
+        WaitForSingleObject.restype = wintypes.DWORD
+    else:
+        GetOverlappedResultEx = ctypes.windll.kernel32.GetOverlappedResultEx
+        GetOverlappedResultEx.argtypes = [wintypes.HANDLE,
+                                          ctypes.POINTER(OVERLAPPED), LPDWORD,
+                                          wintypes.DWORD, wintypes.BOOL]
+        GetOverlappedResultEx.restype = wintypes.BOOL
```

Comment:
@wez 

Inspired by the fix made into the watchman source code to support overlapped operations on Windows 7, we did the same into pywatchman.
Here is the code:
```
    def _GetOverlappedResultEx(self, pipe, olap, nbytes, millis, alertable ):
        if WINDOWS_8_AND_LATER:
            return GetOverlappedResultEx( pipe, olap, nbytes, millis, alertable )

        # else, we fallback to a custom implementation of GetOverlappedResultEx
        # using GetOverlappedResult. This is the exact same implementation
        # that can be found in the watchman source code (stream_win.c)
        # This way, maintenance should be simplified.
        while True:
            if GetOverlappedResult( pipe, olap, nbytes, False ):
                # Result is available
                return True

            starttick = GetTickCount64()
            sleepexreturncode = SleepEx(millis, alertable)
            if sleepexreturncode == WAIT_IO_COMPLETION:
                SetLastError(WAIT_IO_COMPLETION)
                return False
            endtick = GetTickCount64()

            if millis != INFINITE:
                deltatick = endtick - starttick
                durationmillis = millis - deltatick
                if durationmillis <= 0:
                    # Out of time (We suspect we could get rid of this check
                    # using the return of SleepEx. Documentation says that
                    # the method will return 0 if the timeout has been reached).
                    SetLastError(WAIT_TIMEOUT)
                    return False
```

By debugging, we noticed a few things which deserve discussion.

First:
In the following line of the watchman code:
`millis = (DWORD)((ULONGLONG)millis - (end - start));`
reusing millis to store the calculation result forces it to be unsigned.
Since the timeout detection logic is based on having a negative delta (millis <= 0), the timeout detection won't occur unless `(end - start) == millis`, which would be kind of lucky.
Using long long variable type would be more appropriate. For example:
```
long long timeoutvssleepduration = (long long )((ULONGLONG)millis - (end - start));
if (timeoutvssleepduration <= 0) {…}
```

Second:
We noticed that using `GetTickCount64` is not always working as expected. We noticed several calls to a "timeouting" `SleepEx()` being shorter than the timeout value.
E.g.1: getting `deltatick == 1199` when `millis == 1200`
E.g.2: getting `deltatick == 1490` when `millis == 1500`

This leads to a broken detection of a real timeout.
That would deserve a tolerance if the current strategy is kept, at least in pywatchman (we did not debug in the watchman code)

Third:
We ain’t sure if it’s a python-only behavior (probably, since you would have noticed it before with the watchman service), but we observed that the first read op following a write op timeouts according to the call to `SleepEx`. On the second execution of the `while` loop, the call to  `GetOverlappedResult` always succeeds.

Here are the results of 7 executions of `hg status` in debug mode (we get a similar behavior when running not attached to a debugger.):

```
- - - - - - - clock: write (32 bytes)
- F - T F F F clock: read (13 bytes)
- - - - - - - clock: read (48 bytes)
- - - - - - - query: write (246 bytes)
F F F - T F T query: read (13 bytes)
- - - - - - - query: read (80 bytes)

F = Call failed at the first iteration loop but succeed on the second.
T = Call timed out (The timeout detection worked as expected.
There should be no F, only T… so it make obvious that the strategy to detect a timeout needs to be more robust
```

Our guess to explain the fails is that the read operation completes between the return from `GetOverlappedResult()` and the call to `SleepEx()`. This way `GetOverlappedResult` returns 0 BUT `SleepEx` timeout because the `IOCompletion` callback occurred just before. This is likely to happen because the call to `GetTickCount64()` delays just enough the call to `SleepEx`.
Note that adding a 10ms sleep before the call to `GetOverlappedResult( pipe, olap, nbytes, False )` fixes the problem because it allows the IO operation to complete.


Questions:
Would it be possible to verify this behavior on your side with the watchman service?
We are not setup to make those tests on our side. Maybe the situation doesn’t occur because calling Windows API from native code is faster than calling it from python using ctypes.windll.kernel32.*
To help reproducing the problem, maybe you could add a sleep between the calls to `GetOverlappedResult` and `SleepEx` to simulate a longer processing time of `GetTickCount64`

Conclusion:
While we agree that the solution we provided brings a lot of conditional code, it seems safer this way because of the use of an Event. Maybe the best approach would be to drop our wrapper, but still implement the `GetOverlappedResultEx` for Windows7 using the Event. Then, this improvement could be replicated in the C implementation.